mira/metamodel/decapodes.py

```diff
@@ -0,0 +1,481 @@
+                                       src_proj1)
+                self.find_srcs_for_op2(root_var_id,
+                                       binary_operator_1.res.variable_id,
+                                       binary_operator_1.proj2.variable_id,
+                                       src_proj1)
+            for binary_operator_2 in src_proj2:
+                self.find_srcs_for_op2(root_var_id,
+                                       binary_operator_2.res.variable_id,
+                                       binary_operator_2.proj1.variable_id,
+                                       src_proj2)
+                self.find_srcs_for_op2(root_var_id,
+                                       binary_operator_2.res.variable_id,
+                                       binary_operator_2.proj2.variable_id,
+                                       src_proj2)
+
+    def get_only_outputs_op1(self):
+        inputs = set()
+        for op1 in self.op1s.values():
+            inputs.add(op1.src)
+        return set(self.variables.values()) - inputs
+
+    def get_only_outputs_op2(self):
+        inputs = set()
+        for op2 in self.op2s.values():
+            inputs.add(op2.proj1)
+            inputs.add(op2.proj2)
+        return set(self.variables.values()) - inputs
+
+    def get_only_outputs_both(self):
+        inputs = set()
+        for op1 in self.op1s.values():
+            inputs.add(op1.src)
+        for op2 in self.op2s.values():
+            inputs.add(op2.proj1)
+            inputs.add(op2.proj2)
+        return set(self.variables.values()) - inputs
+
+    def get_only_inputs_op1(self):
+        outputs = set()
+        for op1 in self.op1s.values():
+            outputs.add(op1.tgt)
+        return set(self.variables.values()) - outputs
+
+    def get_only_inputs_op2(self):
+        outputs = set()
+        for op2 in self.op2s.values():
+            outputs.add(op2.res)
+        return set(self.variables.values()) - outputs
+
+    def get_only_inputs_both(self):
+        outputs = set()
+        for op1 in self.op1s.values():
+            outputs.add(op1.tgt)
+        for op2 in self.op2s.values():
+            outputs.add(op2.res)
+        return set(self.variables.values()) - outputs
+
+    def get_op1_targets(self):
+        return {op.tgt: op_id for op_id, op in self.op1s.items()}
+
+    def get_op2_targets(self):
+        return {op.res: op_id for op_id, op in self.op2s.items()}
+
+
+class Variable:
+    def __init__(self, variable_id=None, type=None, name=None,
+                 identifiers=None):
+
+        self.variable_id = variable_id
+        self.type = type
+        self.name = name
+        self.symbol = sympy.Symbol(name)
+        self.expression = None
+        self.identifiers = identifiers
+
+    def __repr__(self):
+        return self.name
+
+    def __str__(self):
+        return self.__repr__()
+
+    def build_expression(self, decapode):
```

Comment:
This, and the other method below could be made methods of Decapode instead, and they could be called in the Decapode constructor so that by the time the Decapode is instantiated we have full expressions for each Variable.